Intertrabecular Angles Wrapper fails in Headless mode

I am attempting to run the intertrabecular angles plugin in headless mode. The image I am running this on is shrew.tif Which can be found at the bottom of this page. 

https://bonej.org/ef

I have enclosed the script I am using below to run the wrapper with comments. 

```
# @CommandService cs  
# @TableIOService tios

from ij import IJ
from ij import ImagePlus
from org.scijava.table.io import TableIOOptions
from org.bonej.utilities import SharedTable as table
 
#Set data input-output paths 
input_dir = "/projects/"
input_image_path = input_dir+"shrew.tif"
outputdir = "/measurements/"

#clear old results from the table
IJ.run("Clear BoneJ results");
 
#open input image as IJ1-style ImagePlus to be compatible with wrapper  
inputImagePlus = IJ.openImage(input_image_path)
#inputImagePlus.show()

#Run Intertrabecular Angles Plugin specifying parameters 

wrapper = cs.run("org.bonej.wrapperPlugins.IntertrabecularAngleWrapper", False, ["inputImage",inputImagePlus, "minimumValence", 3, "maximumValence", 50, "minimumTrabecularLength", 0, "marginCutOff", 10, "iteratePruning", True, "useClusters", True, "printCentroids", True,"printCulledEdgePercentages", False, "showSkeleton", True])

#must call wrapper.get() to populate the result table with data
wrapperInstance = wrapper.get()

#Get and save the skeleton image. Here, you don't have to specify the name of the output.
skeleton= wrapperInstance.getOutput("")
IJ.save(skeleton, outputdir+"skeleton.tif")

#save the result table
tios.save(table.getTable(), outputdir+"ITA.csv", TableIOOptions())
```


This script works when wrapper = cs.run("org.bonej.wrapperPlugins.InterTravbecularAngleWrapper", is set to True. This runs the plugin with the GUI and has no issues. When it is set in False to run headlessly I find that the script will generate the skeleton image and then proceed to fail. I have included the error message below. The code fails on wrapperInstance = wrapper.get(). The initial issue I was having was that a dialog box was popping up to confirm the largest skeleton was being processed. I could not click ok in headless mode which resulted in failure. Since then Michael has removed this yet there is still a failure in the code. 

```
File "New_.py", line 28, in <module>
> 	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
> 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> java.util.concurrent.ExecutionException: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception
> 
> 	at org.python.core.Py.JavaError(Py.java:547)
> 	at org.python.core.Py.JavaError(Py.java:538)
> 	at org.python.core.PyReflectedFunction.__call__(PyReflectedFunction.java:192)
> 	at org.python.core.PyReflectedFunction.__call__(PyReflectedFunction.java:208)
> 	at org.python.core.PyObject.__call__(PyObject.java:461)
> 	at org.python.core.PyObject.__call__(PyObject.java:465)
> 	at org.python.core.PyMethod.__call__(PyMethod.java:126)
> 	at org.python.pycode._pyx11.f$0(New_.py:35)
> 	at org.python.pycode._pyx11.call_function(New_.py)
> 	at org.python.core.PyTableCode.call(PyTableCode.java:173)
> 	at org.python.core.PyCode.call(PyCode.java:18)
> 	at org.python.core.Py.runCode(Py.java:1687)
> 	at org.python.core.__builtin__.eval(__builtin__.java:497)
> 	at org.python.core.__builtin__.eval(__builtin__.java:501)
> 	at org.python.util.PythonInterpreter.eval(PythonInterpreter.java:255)
> 	at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:57)
> 	at org.python.jsr223.PyScriptEngine.eval(PyScriptEngine.java:31)
> 	at javax.script.AbstractScriptEngine.eval(AbstractScriptEngine.java:264)
> 	at org.scijava.script.ScriptModule.run(ScriptModule.java:157)
> 	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:63)
> 	at org.scijava.thread.DefaultThreadService.lambda$wrap$2(DefaultThreadService.java:225)
> 	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
> 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
> 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
> 	at java.lang.Thread.run(Thread.java:748)
> Caused by: java.util.concurrent.ExecutionException: java.lang.RuntimeException: Module threw exception
> 	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
> 	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
> 	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
> 	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
> 	at java.lang.reflect.Method.invoke(Method.java:498)
> 	at org.python.core.PyReflectedFunction.__call__(PyReflectedFunction.java:190)
> 	... 24 more
> Caused by: java.lang.RuntimeException: Module threw exception
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:127)
> 	... 6 more
> Caused by: java.lang.NullPointerException
> 	at sc.fiji.analyzeSkeleton.ita.GraphPruning.length(GraphPruning.java:458)
> 	at sc.fiji.analyzeSkeleton.ita.GraphPruning.euclideanDistance(GraphPruning.java:339)
> 	at sc.fiji.analyzeSkeleton.ita.GraphPruning.lambda$pruneShortEdges$0(GraphPruning.java:155)
> 	at java.util.ArrayList.forEach(ArrayList.java:1257)
> 	at sc.fiji.analyzeSkeleton.ita.GraphPruning.pruneShortEdges(GraphPruning.java:155)
> 	at org.bonej.wrapperPlugins.IntertrabecularAngleWrapper.run(IntertrabecularAngleWrapper.java:216)
> 	at org.scijava.command.CommandModule.run(CommandModule.java:196)
> 	at org.scijava.module.ModuleRunner.run(ModuleRunner.java:163)
> 	at org.scijava.module.ModuleRunner.call(ModuleRunner.java:124)
> 	... 6 more
```

> This script works when wrapper = cs.run("org.bonej.wrapperPlugins.InterTravbecularAngleWrapper", is set to True. This runs the plugin with the GUI and has no issues. When it is set in False to run headlessly I find that the script will generate the skeleton image and then proceed to fail

I am not sure that the second argument is meant as a flag to run headless. The [Javadoc](url) rather states:

> process - If true, executes the command with pre- and postprocessing steps from all available [PreprocessorPlugin](https://javadoc.scijava.org/SciJava/org/scijava/module/process/PreprocessorPlugin.html)s and [PostprocessorPlugin](https://javadoc.scijava.org/SciJava/org/scijava/module/process/PostprocessorPlugin.html)s in the plugin index; if false, executes the command with no pre- or postprocessing. For more fine-grained control, see the [ModuleService.run(org.scijava.module.ModuleInfo, boolean, java.lang.Object...)](https://javadoc.scijava.org/SciJava/org/scijava/module/ModuleService.html#run-org.scijava.module.ModuleInfo-boolean-java.lang.Object...-) methods.

I think what is happening is that there is a `NullPointerException` caused by an uninitialised instance variable, which for reasons I don't fully understand, occurs only when called from a script as you have here.

```
> Caused by: java.lang.NullPointerException
> 	at sc.fiji.analyzeSkeleton.ita.GraphPruning.length(GraphPruning.java:458)
```
[This commit](https://github.com/bonej-org/BoneJ2/commit/761b3ca4b7255e397eeb56544ee4353afbdafea0) should fix it.

--- a/Modern/wrapperPlugins/src/main/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapper.java
+++ b/Modern/wrapperPlugins/src/main/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapper.java
@@ -137,14 +137,14 @@ public class IntertrabecularAngleWrapper extends BoneJCommand {
 		style = NumberWidget.SLIDER_STYLE, persistKey = "ITA_max_valence",
 		callback = "enforceValidRange")
 	private int maximumValence = 3;
-	@Parameter(label = "Minimum trabecular length (px)", min = "0",
-		stepSize = "1",
-		description = "Minimum length for a trabecula to be kept from being fused into a node",
-		style = NumberWidget.SPINNER_STYLE, callback = "calculateRealLength",
-		persist = false, initializer = "initRealLength")
-	private int minimumTrabecularLength;
-	@Parameter(label = "Margin (px)", min = "0", stepSize = "1",
-		description = "Nodes with centroids closer than this value to any image boundary will not be included in results",
+	@Parameter(label = "Minimum trabecular length (real units)", min = "0.00",
+		stepSize = "0.1",
+		description = "Minimum length in calibrated units for a trabecula to be kept from being fused into a node",
+		style = NumberWidget.SPINNER_STYLE,
+		persist = false)
+	private double minimumTrabecularLength;
+	@Parameter(label = "Margin (pixels)", min = "0", stepSize = "1",
+		description = "Nodes with centroids closer than this value in pixels to any image boundary will not be included in results",
 		style = NumberWidget.SPINNER_STYLE)
 	private int marginCutOff;
 	@Parameter(label = "Iterate pruning",

--- a/Modern/wrapperPlugins/src/main/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapper.java
+++ b/Modern/wrapperPlugins/src/main/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapper.java
@@ -77,7 +77,6 @@ import net.imglib2.util.ValuePair;
 import org.bonej.utilities.ImagePlusUtil;
 import org.bonej.utilities.SharedTable;
 import org.bonej.wrapperPlugins.wrapperUtils.Common;
-import org.bonej.wrapperPlugins.wrapperUtils.ResultUtils;
 import org.joml.Vector3d;
 import org.scijava.ItemIO;
 import org.scijava.app.StatusService;
@@ -148,8 +147,6 @@ public class IntertrabecularAngleWrapper extends BoneJCommand {
 		description = "Nodes with centroids closer than this value to any image boundary will not be included in results",
 		style = NumberWidget.SPINNER_STYLE)
 	private int marginCutOff;
-	@SuppressWarnings("unused")
-	private String realLength = "";
 	@Parameter(label = "Iterate pruning",
 		description = "If true, iterate pruning as long as short edges remain, or stop after a single pass",
 		required = false, persistKey = "ITA_iterate")
@@ -190,12 +187,13 @@ public class IntertrabecularAngleWrapper extends BoneJCommand {
 	@Parameter
 	private LogService logService;
 
-	private double[] coefficients;
+	private double[] pixelSpacing;
 	private boolean anisotropyWarned;
 
 	@Override
 	public void run() {
 		statusService.showStatus("Intertrabecular angles: Initialising...");
+		initRealLength();
 		statusService.showStatus("Intertrabecular angles: skeletonising");
 		final ImagePlus skeleton = skeletonise();
 		if (showSkeleton)
@@ -214,7 +212,7 @@ public class IntertrabecularAngleWrapper extends BoneJCommand {
 		statusService.showProgress(1, PROGRESS_STEPS);
 		final ValuePair<Graph, double[]> pruningResult = GraphPruning
 			.pruneShortEdges(largestGraph, minimumTrabecularLength, iteratePruning,
-				useClusters, coefficients);
+				useClusters, pixelSpacing);
 		final Graph cleanGraph = pruningResult.a;
 		statusService.showStatus(
 			"Intertrabecular angles: valence sorting trabeculae");
@@ -248,15 +246,6 @@ public class IntertrabecularAngleWrapper extends BoneJCommand {
 		analyser.setup("", skeleton);
 		analyser.run();
 		return analyser.getGraphs();
-	}
-
-	@SuppressWarnings("unused")
-	private void calculateRealLength() {
-		final double calibratedMinimumLength = minimumTrabecularLength *
-			coefficients[0];
-		final String unit = ResultUtils.getUnitHeader(inputImage);
-		realLength = String.join(" ", String.format("%.2g",
-			calibratedMinimumLength), unit);
 	}
 
 	private Map<Integer, DoubleStream> createRadianMap(
@@ -316,21 +305,17 @@ public class IntertrabecularAngleWrapper extends BoneJCommand {
 		}
 	}
 
-	@SuppressWarnings("unused")
 	private void initRealLength() {
 		if (inputImage == null || inputImage.getCalibration() == null) {
-			coefficients = new double[] { 1.0, 1.0, 1.0 };
-			realLength = String.join(" ", String.format("%.2g",
-				(double) minimumTrabecularLength));
+			pixelSpacing = new double[] { 1.0, 1.0, 1.0 };
 		}
 		else {
 			final Calibration calibration = inputImage.getCalibration();
-			coefficients = new double[] { calibration.pixelWidth,
+			pixelSpacing = new double[] { calibration.pixelWidth,
 				calibration.pixelHeight, calibration.pixelDepth };
-			calculateRealLength();
 		}
 	}
-
+	
 	private boolean isCloseToBoundary(final Vertex v) {
 		final Vector3d centroid = PointUtils.centroid(v.getPoints());
 		final int width = inputImage.getWidth();

--- a/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapperTest.java
+++ b/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapperTest.java
@@ -47,11 +47,11 @@ import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.timeout;
 import static org.mockito.Mockito.verify;
 import static org.mockito.Mockito.when;
-import static org.scijava.ui.DialogPrompt.MessageType.WARNING_MESSAGE;
 
 import ij.IJ;
 import ij.ImagePlus;
 import ij.gui.NewImage;
+import ij.measure.Calibration;
 
 import java.net.URL;
 import java.util.Iterator;
@@ -80,7 +80,7 @@ import org.scijava.ui.swing.sdi.SwingDialogPrompt;
 public class IntertrabecularAngleWrapperTest extends AbstractWrapperTest {
 
 	@Test
-	public void testAngleResults() throws Exception {
+	public void testAngleResultsUncalibratedPixels() throws Exception {
 		// SETUP
 		final Predicate<Double> nonEmpty = Objects::nonNull;
 		final URL resource = getClass().getClassLoader().getResource(
@@ -115,6 +115,91 @@ public class IntertrabecularAngleWrapperTest extends AbstractWrapperTest {
 		assertEquals(2, fiveColumn.stream().filter(nonEmpty)
 			.filter(d -> d == Math.PI / 2).count());
 	}
+	
+	@Test
+	public void testAngleResultsIsotropicBigPixels() throws Exception {
+		// SETUP
+		final Predicate<Double> nonEmpty = Objects::nonNull;
+		final URL resource = getClass().getClassLoader().getResource(
+			"test-skelly.zip");
+		assert resource != null;
+		final ImagePlus skelly = IJ.openImage(resource.getFile());
+		Calibration cal = new Calibration();
+		cal.pixelDepth = 2;
+		cal.pixelHeight = 2;
+		cal.pixelWidth = 2;
+		skelly.setCalibration(cal);
+
+		// EXECUTE
+		final CommandModule module = command().run(
+			IntertrabecularAngleWrapper.class, true, "inputImage", skelly,
+			"minimumValence", 3, "maximumValence", 50, "minimumTrabecularLength", 4,
+			"marginCutOff", 0, "useClusters", true, "iteratePruning", false).get();
+
+		// VERIFY
+		@SuppressWarnings("unchecked")
+		final List<DefaultColumn<Double>> table =
+			(List<DefaultColumn<Double>>) module.getOutput("resultsTable");
+		assertNotNull(table);
+		assertEquals(2, table.size());
+		final DefaultColumn<Double> threeColumn = table.get(0);
+		assertEquals("3", threeColumn.getHeader());
+		assertEquals(10, threeColumn.size());
+		assertEquals(3, threeColumn.stream().filter(nonEmpty).count());
+		assertEquals(2, threeColumn.stream().filter(nonEmpty).distinct().count());
+		final DefaultColumn<Double> fiveColumn = table.get(1);
+		assertEquals("5", fiveColumn.getHeader());
+		assertEquals(10, fiveColumn.size());
+		assertEquals(10, fiveColumn.stream().filter(nonEmpty).count());
+		assertEquals(6, fiveColumn.stream().filter(nonEmpty).distinct().count());
+		assertEquals(1, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == Math.PI).count());
+		assertEquals(2, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == Math.PI / 2).count());
+	}
+	
+	@Test
+	public void testAngleResultsIsotropicSmallPixels() throws Exception {
+		// SETUP
+		final Predicate<Double> nonEmpty = Objects::nonNull;
+		final URL resource = getClass().getClassLoader().getResource(
+			"test-skelly.zip");
+		assert resource != null;
+		final ImagePlus skelly = IJ.openImage(resource.getFile());
+		Calibration cal = new Calibration();
+		cal.pixelDepth = 0.1; 
+		cal.pixelHeight = 0.1;
+		cal.pixelWidth = 0.1;
+		skelly.setCalibration(cal);
+
+		// EXECUTE
+		final CommandModule module = command().run(
+			IntertrabecularAngleWrapper.class, true, "inputImage", skelly,
+			"minimumValence", 3, "maximumValence", 50, "minimumTrabecularLength", 0.2,
+			"marginCutOff", 0, "useClusters", true, "iteratePruning", false).get();
+
+		// VERIFY
+		@SuppressWarnings("unchecked")
+		final List<DefaultColumn<Double>> table =
+			(List<DefaultColumn<Double>>) module.getOutput("resultsTable");
+		assertNotNull(table);
+		assertEquals(2, table.size());
+		final DefaultColumn<Double> threeColumn = table.get(0);
+		assertEquals("3", threeColumn.getHeader());
+		assertEquals(10, threeColumn.size());
+		assertEquals(3, threeColumn.stream().filter(nonEmpty).count());
+		assertEquals(2, threeColumn.stream().filter(nonEmpty).distinct().count());
+		final DefaultColumn<Double> fiveColumn = table.get(1);
+		assertEquals("5", fiveColumn.getHeader());
+		assertEquals(10, fiveColumn.size());
+		assertEquals(10, fiveColumn.stream().filter(nonEmpty).count());
+		assertEquals(6, fiveColumn.stream().filter(nonEmpty).distinct().count());
+		assertEquals(1, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == Math.PI).count());
+		assertEquals(2, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == Math.PI / 2).count());
+	}
+	
 
 	@Test
 	public void testAnisotropicImageShowsWarningDialog() {

--- a/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapperTest.java
+++ b/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/IntertrabecularAngleWrapperTest.java
@@ -39,14 +39,9 @@ import static org.junit.Assert.assertNull;
 import static org.junit.Assert.assertTrue;
 import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.ArgumentMatchers.anyString;
-import static org.mockito.ArgumentMatchers.eq;
-import static org.mockito.ArgumentMatchers.startsWith;
-import static org.mockito.Mockito.after;
 import static org.mockito.Mockito.doNothing;
-import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.timeout;
 import static org.mockito.Mockito.verify;
-import static org.mockito.Mockito.when;
 
 import ij.IJ;
 import ij.ImagePlus;
@@ -67,7 +62,6 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.scijava.command.CommandModule;
 import org.scijava.table.DefaultColumn;
-import org.scijava.ui.swing.sdi.SwingDialogPrompt;
 
 /**
  * Tests for {@link IntertrabecularAngleWrapper}
@@ -87,7 +81,7 @@ public class IntertrabecularAngleWrapperTest extends AbstractWrapperTest {
 			"test-skelly.zip");
 		assert resource != null;
 		final ImagePlus skelly = IJ.openImage(resource.getFile());
-
+		
 		// EXECUTE
 		final CommandModule module = command().run(
 			IntertrabecularAngleWrapper.class, true, "inputImage", skelly,
@@ -200,7 +194,6 @@ public class IntertrabecularAngleWrapperTest extends AbstractWrapperTest {
 			.filter(d -> d == Math.PI / 2).count());
 	}
 	
-
 	@Test
 	public void testAnisotropicImageShowsWarningDialog() {
 		CommonWrapperTests.testAnisotropyWarning(imageJ(),

--- a/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/headless/AbstractWrapperHeadlessTest.java
+++ b/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/headless/AbstractWrapperHeadlessTest.java
@@ -1,0 +1,105 @@
+/*-
+ * #%L
+ * High-level BoneJ2 commands.
+ * %%
+ * Copyright (C) 2015 - 2020 Michael Doube, BoneJ developers
+ * %%
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ * 
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ * 
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ * #L%
+ */
+/*
+BSD 2-Clause License
+Copyright (c) 2020, Michael Doube, Richard Domander, Alessandro Felder
+All rights reserved.
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions are met:
+* Redistributions of source code must retain the above copyright notice, this
+  list of conditions and the following disclaimer.
+* Redistributions in binary form must reproduce the above copyright notice,
+  this list of conditions and the following disclaimer in the documentation
+  and/or other materials provided with the distribution.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+package org.bonej.wrapperPlugins.headless;
+
+import net.imagej.ImageJ;
+import org.bonej.utilities.SharedTable;
+import org.junit.After;
+import org.junit.AfterClass;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.scijava.command.CommandService;
+
+/**
+ * An abstract base test class that handles basic setup and tear down
+ * for testing wrapper plugins in headless mode.
+ * <p>
+ * NB Remember to call the methods in this class if you override/hide the methods!
+ * </p>
+ * @author Richard Domander
+ * @author Michael Doube
+ */
+public abstract class AbstractWrapperHeadlessTest {
+    private static ImageJ imageJ;
+    private static CommandService commandService;
+
+    protected static CommandService command() {
+        return commandService;
+    }
+
+    protected static ImageJ imageJ() {
+        return imageJ;
+    }
+
+    @BeforeClass
+    public static void basicOneTimeSetup() {
+        imageJ = new ImageJ();
+        commandService = imageJ.command();
+    }
+
+    @Before
+    public void setup() {
+        imageJ.ui().setHeadless(true);
+    }
+
+    @After
+    public void tearDown() {
+        SharedTable.reset();
+    }
+
+    @AfterClass
+    public static void basicOneTimeTearDown() {
+        imageJ.context().dispose();
+        imageJ = null;
+        commandService = null;
+    }
+}

--- a/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/headless/AbstractWrapperHeadlessTest.java
+++ b/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/headless/AbstractWrapperHeadlessTest.java
@@ -26,28 +26,6 @@
  * POSSIBILITY OF SUCH DAMAGE.
  * #L%
  */
-/*
-BSD 2-Clause License
-Copyright (c) 2020, Michael Doube, Richard Domander, Alessandro Felder
-All rights reserved.
-Redistribution and use in source and binary forms, with or without
-modification, are permitted provided that the following conditions are met:
-* Redistributions of source code must retain the above copyright notice, this
-  list of conditions and the following disclaimer.
-* Redistributions in binary form must reproduce the above copyright notice,
-  this list of conditions and the following disclaimer in the documentation
-  and/or other materials provided with the distribution.
-THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
-AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
-IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
-DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
-FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
-DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
-SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
-CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
-OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
-*/
 
 package org.bonej.wrapperPlugins.headless;
 

--- a/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/headless/IntertrabecularAnglesWrapperHeadlessTest.java
+++ b/Modern/wrapperPlugins/src/test/java/org/bonej/wrapperPlugins/headless/IntertrabecularAnglesWrapperHeadlessTest.java
@@ -1,0 +1,134 @@
+package org.bonej.wrapperPlugins.headless;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertTrue;
+
+import java.net.URL;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Predicate;
+
+import net.imagej.table.DefaultResultsTable;
+
+import org.bonej.wrapperPlugins.IntertrabecularAngleWrapper;
+import org.junit.Test;
+import org.scijava.command.CommandModule;
+import org.scijava.table.DefaultColumn;
+import org.scijava.table.DefaultGenericTable;
+import org.scijava.table.DoubleColumn;
+
+import ij.IJ;
+import ij.ImagePlus;
+import ij.measure.Calibration;
+
+/**
+ * Tests for BoneJ Wrapper Plugins that do not rely on a GUI.
+ * 
+ * Its intention is to validate output when plugin wrappers are called
+ * from scripts and macros.
+ * 
+ * @author Michael Doube
+ *
+ */
+public class IntertrabecularAnglesWrapperHeadlessTest extends AbstractWrapperHeadlessTest {
+
+	@Test
+	public void testInterTrabecularAngles() throws InterruptedException, ExecutionException {
+		//SETUP
+		assertTrue(imageJ().ui().isHeadless());
+		final Predicate<Double> nonEmpty = Objects::nonNull;
+
+		final URL resource = getClass().getClassLoader().getResource(
+				"test-skelly.zip");
+		assert resource != null;
+		final ImagePlus skelly = IJ.openImage(resource.getFile());
+		Calibration cal = new Calibration();
+		cal.pixelDepth = 0.1; 
+		cal.pixelHeight = 0.1;
+		cal.pixelWidth = 0.1;
+		skelly.setCalibration(cal);
+
+		//EXECUTE
+		final CommandModule module = command().run(
+			IntertrabecularAngleWrapper.class, false, "inputImage", skelly,
+			"minimumValence", 3, "maximumValence", 50, "minimumTrabecularLength", 0.2,
+			"marginCutOff", 0, "useClusters", true, "printCentroids", true,
+			"iteratePruning", false, "showSkeleton", true).get();
+
+		Map<String,Object> outputs = module.getOutputs();
+
+		//VERIFY
+		assertNotNull(outputs);
+		logOutputNameAndClass(module);
+		
+		//check the output image exists and is an ImagePlus
+		assertNotNull(module.getOutput("skeletonImage"));
+		assertTrue(outputs.get("skeletonImage") instanceof ImagePlus);
+
+		//check that the angle table exists and is a DefaultGenericTable
+		assertNotNull(module.getOutput("resultsTable"));
+		assertTrue(module.getOutput("resultsTable") instanceof DefaultGenericTable);
+
+		//Check the angle table contains the expected values.
+		final DefaultGenericTable table = (DefaultGenericTable) module.getOutput("resultsTable");
+		assertEquals(2, table.size());
+		@SuppressWarnings("unchecked")
+		final DefaultColumn<Double> threeColumn = (DefaultColumn<Double>) table.get(0);
+		assertEquals("3", threeColumn.getHeader());
+		assertEquals(10, threeColumn.size());
+		assertEquals(3, threeColumn.stream().filter(nonEmpty).count());
+		assertEquals(2, threeColumn.stream().filter(nonEmpty).distinct().count());
+		@SuppressWarnings("unchecked")
+		final DefaultColumn<Double> fiveColumn = (DefaultColumn<Double>) table.get(1);
+		assertEquals("5", fiveColumn.getHeader());
+		assertEquals(10, fiveColumn.size());
+		assertEquals(10, fiveColumn.stream().filter(nonEmpty).count());
+		assertEquals(6, fiveColumn.stream().filter(nonEmpty).distinct().count());
+		assertEquals(1, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == Math.PI).count());
+		assertEquals(2, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == Math.PI / 2).count());
+		assertEquals(2, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == 0.8621700546672264).count());
+		assertEquals(2, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == 2.279422598922567).count());
+		assertEquals(2, fiveColumn.stream().filter(nonEmpty)
+			.filter(d -> d == 2.4329663814621227).count());
+		
+		//check that the centroid table exists and is a DefaulResultsTable
+		assertNotNull(module.getOutput("centroidTable"));
+		assertTrue(module.getOutput("centroidTable") instanceof DefaultResultsTable);
+		
+		//check that the centroid table contains the expected values
+		final DefaultResultsTable centroidTable = (DefaultResultsTable) module.getOutput("centroidTable");
+		assertEquals(6, centroidTable.size());
+		assertEquals(7, centroidTable.getRowCount());
+		final DoubleColumn v1x = centroidTable.get(0);
+		assertEquals(2, v1x.stream().filter(d -> d == 2).count());
+		assertEquals(1, v1x.stream().filter(d -> d == 9).count());
+		assertEquals(2, v1x.stream().filter(d -> d == 17).count());
+		assertEquals(2, v1x.stream().filter(d -> d == 24).count());
+	}
+	
+	/**
+	 * Print the names and object types in the output list to the console
+	 *  
+	 * @param module
+	 */
+	private void logOutputNameAndClass(CommandModule module) {
+		
+		Map<String,Object> outputs = module.getOutputs();
+		
+		System.out.println("Output list for " + module.getInfo().getClassName());
+		
+		for (Map.Entry<String, Object> entry : outputs.entrySet()) {
+			if (Objects.isNull(entry))
+				continue;
+			if (Objects.isNull(entry.getKey()) || Objects.isNull(entry.getValue()))
+				continue;
+			System.out.println(entry.getKey() + " " + entry.getValue().getClass().getName());
+		}
+	}
+}

--- a/README.md
+++ b/README.md
@@ -7,7 +7,7 @@ BoneJ is a collection of Fiji/ImageJ plug-ins for skeletal biology. It provides 
 > doi:[10.12688/wellcomeopenres.16619.1](https://doi.org/10.12688/wellcomeopenres.16619.1)
 
 ## Links
-* [User guide](https://imagej.github.io/plugins/bonej)
+* [User guide](https://imagej.net/plugins/bonej)
 * [ImageJ forum](https://forum.image.sc/tags/bonej)
 * [Developer documentation](https://github.com/bonej-org/BoneJ2/wiki)
 

